Implement a service supervisor for Postgres.

DimCitus opened the pull request:
At the moment the postgres controller process is not used in the main code
of pg_autoctl, the idea is to make use of it in later PRs implementing the
monitor and then the keeper process supervision and integration.

To use this:

```
$ export PGDATA=...
$ pg_autoctl create postgres ...
$ pg_ctl stop
$ PG_AUTOCTL_DEBUG=1 ./src/bin/pg_autoctl/pg_autoctl do service postgres
$ PG_AUTOCTL_DEBUG=1 ./src/bin/pg_autoctl/pg_autoctl do pgctl on
$ PG_AUTOCTL_DEBUG=1 ./src/bin/pg_autoctl/pg_autoctl do pgctl off
```

JelteF commented:
When running `do pgctl on` it should not say "... with pid 0". Either show the actual pid or do not show it at all.

```bash
PG_AUTOCTL_DEBUG=true pg_autoctl do pgctl on --pgdata node1
16:11:40 24878 INFO  Postgres is serving PGDATA "/home/jelte/work/pg_auto_failover/node1" on port 5433 with pid 0
```

DimCitus commented:
> When running `do pgctl on` it should not say "... with pid 0". Either show the actual pid or do not show it at all.
> 
> ```shell
> PG_AUTOCTL_DEBUG=true pg_autoctl do pgctl on --pgdata node1
> 16:11:40 24878 INFO  Postgres is serving PGDATA "/home/jelte/work/pg_auto_failover/node1" on port 5433 with pid 0
> ```

Thanks, fixed now. I've been using the “wrong” pgSetup instance in the log output. Simplified/fixed by using a single pgSetup instance, so there can't be confusion.

DimCitus commented on src/bin/pg_autoctl/supervisor.c:
```diff
@@ -0,0 +1,646 @@
+						log_info("Waiting for subprocesses to terminate.");
+					}
+
+					/*
+					 * If we've been waiting for quite a while for
+					 * sub-processes to terminate, it might be because a signal
+					 * was sent to only the process leader, such as when doing
+					 * `pkill pg_autoctl`, rather than to the process group,
+					 * such as when using killpg(2) or `pg_autoctl stop`.
+					 *
+					 * In that case let's signal all our process group
+					 * ourselves and see what happens next.
+					 */
```
That looks like an outdated comment from before I switched to forwarding the signal explicitly. Thanks for the comment, I will remove/update this comment!

DimCitus replied:
Comment edited.

DimCitus commented on src/bin/lib/subcommands.c/runprogram.h:
```diff
@@ -196,14 +215,27 @@ execute_program(Program *prog)
 			int stdIn = open(DEV_NULL, O_RDONLY);
```
Added.

JelteF replied:
You didn't actually change this call site.

DimCitus replied:
Oops, good point, done now.

DimCitus commented on src/bin/lib/subcommands.c/runprogram.h:
```diff
@@ -236,13 +268,79 @@ execute_program(Program *prog)
+
+	/* Flush stdio channels just before fork, to avoid double-output problems */
+	fflush(stdout);
+	fflush(stderr);
```
Fixed the comment, I believe we still need the fflush calls.

DimCitus commented on src/bin/lib/subcommands.c/runprogram.h:
```diff
@@ -236,13 +268,79 @@ execute_program(Program *prog)
+	 * apparently closing it may cause undefined behavior if any read was to
+	 * happen.
+	 */
+	int stdIn = open(DEV_NULL, O_RDONLY);
```
Done now.

DimCitus commented on src/bin/pg_autoctl/pgctl.c:
```diff
@@ -919,6 +982,97 @@ pg_ctl_start(const char *pg_ctl,
+ * pg_ctl_postmaster runs the "postmaster" command-line in the current process,
+ * with the same options as we would use in pg_ctl_start. pg_ctl_postmaster
+ * does not fork a Postgres process in the background, we keep the control over
+ * the postmaster process. Think exec() rather then fork().
```
Added.

DimCitus commented on src/bin/lib/subcommands.c/runprogram.h:
```diff
@@ -236,13 +268,79 @@ execute_program(Program *prog)
 
 
+/*
+ * Run given program with its args, by using exec().
```
Added.

DimCitus commented on src/bin/pg_autoctl/state.h:
```diff
@@ -133,7 +146,43 @@ typedef struct
+} ExpectedPostgresStatus;
+
+/*
+ *  Note: This struct is serialized/serialiazed to/from state file. Therefore
```
I think I meant to say serialized/deserialized here, fixed that way.

DimCitus commented on src/bin/pg_autoctl/supervisor.h:
```diff
@@ -0,0 +1,51 @@
+ * reload functions get passed a context that is a pointer to the Service
+ * struct definition of the service being asked to stop.
+ */
+typedef bool (*ServiceStartFunction)(void *context, pid_t *pid);
+typedef bool (*ServiceStopFunction)(void *context);
+typedef void (*ServiceReloadFunction)(void *context);
```
Oh that's a nice trick, I changed the code to use that instead. Thanks!

DimCitus commented on src/bin/pg_autoctl/supervisor.c:
```diff
@@ -0,0 +1,646 @@
+
+					if (stoppingLoopCounter == 1)
+					{
+						log_info("Waiting for subprocesses to terminate.");
```
I think the intention was that with some luck we proceed on the first loop to find exited processes and then we will never have to display the "waiting" message to the user. It would be too fast for that...

JelteF replied:
If that's the intention then it doesn't do that, since the `stoppingLoopCounter` is incremented right before this check. So it's effectively the same check as `stoppingLoopCounter == 0` before the increment (that's what I tried to indicate with my previous comment, but that was not very clear).

I'm not sure if the intention you mention is indeed what we want, since we will start waiting as soon as we call `supervisor_stop_process`. But if you disagree I'm fine with that.

DimCitus replied:
Ah yeah I see what you mean, and feel a little stupid about that. With the refactoring it should now be okay, unless I had another brain fart.

DimCitus commented on src/bin/pg_autoctl/supervisor.c:
```diff
@@ -0,0 +1,646 @@
+					 */
+					if (stoppingLoopCounter == 0)
+					{
+						(void) supervisor_stop_subprocesses(services,
+															subprocessCount);
```
I think I see what you mean, and I'm not sure it would be of much use. Will have to look at the source again with the other PRs to see if we would then be able to have the same function everywhere just calling the new API, but I think we have different local variables specific to each service that we want to use in each place anyway.

DimCitus replied:
The idea of `asked_to_stop_fast` is just that we don't finish the current keeper loop and quit as soon as possible, and `asked_to_stop` allows for finishing current business first. I believe we still have that difference in the way the main keeper look reacts to signals. That might be the only place where we process those two signal variables differently too.

JelteF replied:
> The idea of asked_to_stop_fast is just that we don't finish the current keeper loop and quit as soon as possible, and asked_to_stop allows for finishing current business first.

Without `stopFastFunction` this would not be possible then. Since the supervisor process will swallow the signal and send the same signal to the keeper process no matter what.

DimCitus replied:
The refactoring I just completed now relays the `asked_to_stop_fast` signal to sub-processes:

``` c
	int signal = asked_to_stop_fast ? SIGINT : SIGTERM;
```

DimCitus commented on src/bin/pg_autoctl/supervisor.c:
```diff
@@ -0,0 +1,646 @@
+				 * try to restart the service up to 5 times or 20s, whichever
+				 * comes first.
+				 */
+				else
```
Good point, done that way now.

DimCitus commented on src/bin/pg_autoctl/supervisor.c:
```diff
@@ -0,0 +1,646 @@
+/*
+ * supervisor_reset_service_restart_counters loops over known services and
+ * reset the retries count and stopTime of services that have been known
+ * running for more than a minute (SUPERVISOR_SERVICE_RUNNING_TIME is 60s).
```
Included in the fix commit, thanks.

DimCitus commented on src/bin/pg_autoctl/supervisor.c:
```diff
@@ -0,0 +1,646 @@
+					dead->stopTime = now;
+				}
+
+				/*
+				 * One child process has terminated. If it terminated and
+				 * returned 0 as its exit status, then we consider that our job
+				 * is done here and cleanly exit.
+				 *
+				 * Before we exit though, we need to stop the other running
+				 * services.
+				 */
+				if (returnCode == 0)
+				{
```
If Postgres exits happily, then we do that too, and then systemd (or the user, etc) needs to decide what's next. Systemd will just launch our service again. So yeah that's intended.

DimCitus commented on src/bin/pg_autoctl/supervisor.c:
```diff
@@ -0,0 +1,646 @@
+ *
+ * There is no magic or heuristic that can help us decide if a failure is
+ * transient or permanent, so we implement the simple thing: we restart our
+ * dead service up to 5 times in a row, and spend up to 10 seconds or retrying,
```
Fixed.

DimCitus commented on src/bin/pg_autoctl/supervisor.c:
```diff
@@ -0,0 +1,646 @@
+ * This strategy is inspired by http://erlang.org/doc/man/supervisor.html and
+ * http://erlang.org/doc/design_principles/sup_princ.html#maximum-restart-intensity
+ *
+ *    If more than MaxR number of restarts occur in the last MaxT seconds, the
+ *    supervisor terminates all the child processes and then itself. The
+ *    termination reason for the supervisor itself in that case will be
+ *    shutdown.
```
I'm not sure that the code is that different from the comment here. I think our implementation with the stopTime tracking of the service allows to track that we have restarted the service more than MaxR times in the last MaxT seconds... It's not obvious to me that the two ways to explain things are that different?

JelteF replied:
So, the way I read it.
We do: 
- If we restart 5 times without one process staying alive for 15 seconds, we exit.
- If the process has died and we are still trying to restart after 10 seconds, we exit. NOTE: This can happen at the second restart already, if the first restart failed after 11 seconds. We don't count it as successful start in that case, but we have been restarting for more than 10 seconds, so we exit.

Erlang does:
- Restart whenever the process stops. If this has happened more than MaxR in MaxT seconds already, exit instead. Count the number of restarts in the last MaxT seconds. If this is more than MaxR. Exit. NOTE: This logic does not have a notion of "service running time", this is contained in the MaxT. 


That's why I say they are very different. they have the same variables but use them quite differently. 
Personally I like the Erlang logic, as it seems pretty simple and never considers 2 failures as an exit condition.

DimCitus replied:
I have used the 15s of running heuristic to implement MaxR/MaxT without having to track all the restart times, that way I don't have to count how many restarts actually happened in the last MaxT interval... but maybe the difference in the approach is then such as to be user visible and harder to explain.

DimCitus replied:
Okay, so I have now fixed it with using a ring buffer to keep track of the latest MaxR + 1 restarts. I have left the default values of 5 MaxR in 10 MaxT as the only time this mechanism kicked in for me in development was a startup problem and pg_autoctl would try again and again very quickly, but now only up to 5 times before giving up.

That said we could now say 5 times in  10 mins and this case would be handled properly too. Because we have large timeouts for some operations, such as contacting the monitor when it's not responsive (yet), we might want to change to something like MaxR=5 and MaxT=300 or something in that order of magnitude, what do you think?

JelteF replied:
I think it makes sense to set MaxT to something higher than 10 seconds indeed. I think 300 seconds sounds reasonable.

DimCitus commented on src/bin/pg_autoctl/cli_do_service.c:
```diff
@@ -0,0 +1,176 @@
+	/* display a user-friendly process name */
+	(void) set_ps_title("pg_autoctl: start/stop postgres");
+
+	(void) service_postgres_ctl_loop(&postgres);
```
When using the supervisor approach, we will then actually call `pg_autoctl do service postgres` as a subprocess, this one is not supposed to be yet another supervisor... so I don't think we can exercise the supervisor code much yet in this PR unfortunately...

DimCitus commented on src/bin/pg_autoctl/service_postgres_ctl.c:
```diff
@@ -0,0 +1,511 @@
+	fflush(stderr);
+
+	/* time to create the node_active sub-process */
+	fpid = fork();
```
I think I see what you mean, and I'm not sure it would be of much use. Will have to look at the source again with the other PRs to see if we would then be able to have the same function everywhere just calling the new API, but I think we have different local variables specific to each service that we want to use in each place anyway.

DimCitus commented on src/bin/pg_autoctl/service_postgres_ctl.c:
```diff
@@ -0,0 +1,511 @@
+		}
+
+		/* that's expected the shutdown sequence from the supervisor */
+		if (asked_to_stop || asked_to_stop_fast)
```
Well Postgres stop modes aren't as useful as they look, I think it's okay to always use fast. At least we should never use immediate, of course, because we don't want to lose data. And the smart mode is all but.

JelteF replied:
Okay that makes sense. Especially since with HA, you should be able to handle retries of queries.

DimCitus commented on src/bin/pg_autoctl/pgctl.c:
```diff
@@ -919,6 +982,97 @@ pg_ctl_start(const char *pg_ctl,
+			return false;
+		}
+
+		/* pg_ctl --options can be specified multiple times */
```
Yeah that comment was copy/pasted and forgotten and should have been removed. Done now.

DimCitus commented on src/bin/pg_autoctl/pgctl.c:
```diff
@@ -919,6 +982,97 @@ pg_ctl_start(const char *pg_ctl,
+
+	args[argsIndex] = NULL;
+
+	/* we do not want to call setsid() when running this program. */
```
Added comments. We want a single session leader process and that's the top `pg_autoctl` process.

DimCitus commented on src/bin/pg_autoctl/pgctl.c:
```diff
@@ -919,6 +982,97 @@ pg_ctl_start(const char *pg_ctl,
 
+/*
+ * pg_ctl_postmaster runs the "postmaster" command-line in the current process,
+ * with the same options as we would use in pg_ctl_start. pg_ctl_postmaster
```
Done.

DimCitus commented:
> I undrestand that taking ownership would require killing postgres (or exiting saying postgres is already running and let the user handle that). I'm not sure what we want here, but it's good to at least make a concious decision either way.

On the monitor, it's easy, we want full control. There's now only one way to create a monitor, and that includes creating the Postgres cluster `pg_ctl init`, and then owning the process, so we're good as of the next PR.

On the keeper, we want full control too, but we might still want to allow our users to `pg_autoctl create postgres` on top of an already running Postgres instance. And in that case, we certainly want to avoid doing a restart as the first step of installing a fault tolerance mechanism.

> 1. I think we need to document this process architecture somewhere. This can be done after this PR is merged and do it for the monitor PR, since then it's actually useful. But I think it's good to document all these different processes, what their goals are and how they are created, i.e. by actually running the command.

Good point. I started a README.md file in the `src/bin/pg_autoctl` directory, with light coverage of the point here. Let's see how far we go from there. We might want to refine the document in the next PRs.

> 2. Some of the comments suggested that you should use `killpg` instead of `kill` on the main process. I think as a user running `kill` is way more common, so I think we should make handling that the default (it seems like we already do that).

There were a point in the development of this PR where it would only work when killing the process group, and I had to change `pg_autoctl stop` for that. I have fixed that situation though, and this comment is a left over. Cleaned-up now.

JelteF commented on src/bin/pg_autoctl/pgctl.c:
```diff
@@ -983,10 +983,15 @@ pg_ctl_start(const char *pg_ctl,
 /*
- * pg_ctl_postmaster runs the "postmaster" command-line in the current process,
+ * pg_ctl_postgres runs the "postgres" command-line in the current process,
  * with the same options as we would use in pg_ctl_start. pg_ctl_postmaster
```
```suggestion
 * with the same options as we would use in pg_ctl_start. pg_ctl_postgres
```

DimCitus replied:
Done now.

JelteF commented on src/bin/pg_autoctl/supervisor.h:
```diff
@@ -0,0 +1,51 @@
+ * reload functions get passed a context that is a pointer to the Service
+ * struct definition of the service being asked to stop.
+ */
+typedef bool (*ServiceStartFunction)(void *context, pid_t *pid);
+typedef bool (*ServiceStopFunction)(void *context);
+typedef void (*ServiceReloadFunction)(void *context);
```
As far as I can tell the typedefs are not needed anymore now for the bottom two for consistency sake I think it's nicer to also move the startFunction inline into the struct.
```suggestion
```

DimCitus replied:
Done now. Thanks again!

JelteF commented on src/bin/pg_autoctl/supervisor.c:
```diff
@@ -0,0 +1,651 @@
+							log_warn(
+								"Still waiting for subprocesses to terminate.");
+						}
+					}
```
Good that you updated the comment, but I still think this code is pretty strange in general. The main problems I have in general:
1. If the user asked for a normal shutdown, at some point we will automatically upgrade it to a fast shutdown.
2. We assume these signals work as we expect on the child processes. Which is not necessarily the case, SIGINT, SIGTERM and SIGQUIT have different meaninings on postgres and us.
3. These timeouts cannot be configured.

This is why I think it's better to simply require the user to send a different signal after a different timeout, if the process has not quit yet. I believe this is configurable in both systemd and docker. This does require us to handle an update to the signal, i.e. if the user first sends SIGTERM and then SIGQUIT we still handle the SIGQUIT. Which I think is good to do anyway.

DimCitus replied:
I think the current way this is done is now better. We still handle our own shutdown sequence in case our services won't terminate orderly, but we don't translate a SIGINT into a SIGTERM anymore and just relay the user's signal, which makes us compatible again with external process supervision systems.

JelteF commented on src/bin/pg_autoctl/service_postgres_ctl.c:
```diff
@@ -0,0 +1,522 @@
+ * service_postgres_stop stops the postgres service, using pg_ctl stop.
+ */
+bool
+service_postgres_ctl_stop(void *context)
```
Dead code

DimCitus replied:
Removed.

JelteF commented on src/bin/pg_autoctl/supervisor.h:
```diff
@@ -0,0 +1,112 @@
+ * configuration is wrong or the data directory damaged beyond repair or for
+ * some reasons, the service can't be restarted.
+ *
+ * There is no magic or heuristic that can help us decide if a failure is
+ * transient or permanent, so we implement the simple thing: we restart our
+ * dead service up to 5 times in a row, and spend up to 10 seconds retrying,
+ * and stop as soon as one of those conditions is reached.
```
This comment still explains the old behaviour.

DimCitus replied:
Removed. The per-service Restart Policy is detailed just before with the enum data type introduction so I don't think we need to explain it here again.

JelteF commented on src/bin/pg_autoctl/supervisor.c:
```diff
@@ -724,8 +651,77 @@ supervisor_restart_service(Supervisor *supervisor, Service *service, int status)
+ * of time.
+ */
+static bool
+supervisor_may_restart(Service *service)
```
I think the code to could be a bit simpler (no loop necessary afaict). See pseudo code below:
Initially: 
```c
next_position_to_insert = 0
startTime = empty array of size 5 (not 6 like now)
count = 0 // number of entries in the ringbuffer 
```
Every time a restart happens
```c
// If this is not the case 5 restarts have not yet happend so we can try again for sure.
if (count == 5) {
    
    if (now - starTime[next_position_to_insert] <= 15) {
         
        return false;
    }
}

starTime[next_position_to_insert] = now
if (count < 5) {
    count += 1
}
next_position_to_insert = (next_position_to_insert + 1) % 5
return true
```

DimCitus replied:
Done now, thanks for your input. I think we're back to the previous coding, but with all the intermediate restart time stamps rather than some smarts about keeping only one of them and a weird third variable and reset mechanism. Overall I like the new code better, thanks!

JelteF commented on src/bin/pg_autoctl/supervisor.c:
```diff
@@ -0,0 +1,827 @@
+
+		if (kill(service->pid, SIGHUP) != 0)
+		{
+			log_error("Failed to send SIGQUIT to service %s with pid %d",
```
```suggestion
			log_error("Failed to send SIGHUP to service %s with pid %d",
```

DimCitus replied:
Done.

JelteF commented on src/bin/pg_autoctl/supervisor.c:
```diff
@@ -0,0 +1,827 @@
+			{
+				if (kill(service->pid, signal) != 0)
+				{
+					log_error("Failed to send SIGQUIT to service %s with pid %d",
```
Should use `strsignal` not to replace `SIGQUIT` here.

DimCitus replied:
Done.

JelteF commented on src/bin/pg_autoctl/supervisor.c:
```diff
@@ -0,0 +1,827 @@
+
+
+/*
+ * supervisor_signal_process_group sends a signal (SIGQUIT) to our own process
```
```suggestion
 * supervisor_signal_process_group sends a signal to our own process
```

DimCitus replied:
Done.

JelteF commented on src/bin/pg_autoctl/supervisor.c:
```diff
@@ -0,0 +1,827 @@
+						(void) supervisor_stop_subprocesses(supervisor);
+					}
+
+					(void) supervisor_shutdown_sequence(stoppingLoopCounter++);
```
Can we move the `++`  outside of this call for clarity if the increment happens before or after the call?

DimCitus replied:
Move this bit of code and I think it's not ok?

DimCitus commented on src/bin/pg_autoctl/cli_do_service.c:
```diff
@@ -0,0 +1,365 @@
+				 cli_do_service_getpid);
+
+CommandLine service_pgcontroler =
+	make_command("pgcontroler",
```
Done now, thanks.

DimCitus commented on src/bin/pg_autoctl/pgctl.c:
```diff
@@ -752,19 +798,36 @@ log_program_output(Program prog, int outLogLevel, int errorLogLevel)
+	program = run_program(pg_ctl, "initdb",
+						  "--silent",
+						  "--pgdata", pgdata,
+						  "-o", "'--auth=trust'", /* avoid warning message */
```
Done now. Yeah we are relaying warnings and want to minimise them of course.

DimCitus commented on src/bin/pg_autoctl/service_postgres_ctl.c:
```diff
@@ -0,0 +1,516 @@
+							  "see above for details.");
+
+					/* maybe next round will have better luck? */
+					continue;
```
Done now, with same variable names as in `loop.c`.

DimCitus commented on src/bin/pg_autoctl/service_postgres_ctl.c:
```diff
@@ -0,0 +1,516 @@
+							  missingPgdataIsOk,
+							  postgresNotRunningIsOk) &&
+				pg_setup_pgdata_exists(&newPgSetup) &&
+				pg_auto_failover_default_settings_file_exists(&newPgSetup))
```
Because we first create a buffer string with the content of the file and then write to disk in a single `write()` operation, I'm not sure that's a high-risk...

JelteF replied:
I think that's fair, let's not care for now then.

DimCitus commented on src/bin/pg_autoctl/service_postgres_ctl.c:
```diff
@@ -0,0 +1,516 @@
+ * service_postgres_ctl_reload sends a SIGHUP to the controller process.
+ */
+void
+service_postgres_ctl_reload(void *context)
```
Removed.

DimCitus commented on src/bin/pg_autoctl/service_postgres_ctl.c:
```diff
@@ -0,0 +1,516 @@
+ * - using the state file from the FSM directly makes things overall simpler,
+ *   the only drawback is that this is one more place where we need to
+ *   implement the FSM semantics, it's not just the transition functions... it
+ *   was never "just the transition functions" that said.
```
Removed. That's from the very beginning of this PR when I was still playing with other ideas.

DimCitus commented on src/bin/pg_autoctl/pgctl.c:
```diff
@@ -919,6 +982,105 @@ pg_ctl_start(const char *pg_ctl,
+	/* prepare startup.log file in PGDATA */
+	join_path_components(logfile, pgdata, "startup.log");
+
+	args[argsIndex++] = (char *) postgres;
+	args[argsIndex++] = "-D";
+	args[argsIndex++] = (char *) pgdata;
+	args[argsIndex++] = "-p";
+	args[argsIndex++] = (char *) intToString(pgport).strValue;
+
+	if (!IS_EMPTY_STRING_BUFFER(listen_addresses))
+	{
+		args[argsIndex++] = "-h";
+		args[argsIndex++] = (char *) listen_addresses;
+	}
+
+	if (env_exists("PG_REGRESS_SOCK_DIR"))
+	{
+		if (!get_env_copy("PG_REGRESS_SOCK_DIR", env_pg_regress_sock_dir,
+						  MAXPGPATH))
+		{
+			/* errors have already been logged */
+			return false;
+		}
+
+		args[argsIndex++] = "-k";
+		args[argsIndex++] = (char *) env_pg_regress_sock_dir;
+	}
+
+	args[argsIndex] = NULL;
```
I believe the `postgres` command line does not have support for long options equivalent here, or at least they would not be documented at https://www.postgresql.org/docs/current/app-postgres.html (which is the same thing for us as we don't want to rely on non-documented behaviour).

JelteF commented on src/bin/pg_autoctl/supervisor.c:
```diff
@@ -0,0 +1,880 @@
+				}
+
+				/* allow for processing SIGINT again next time it's sent */
+				if (asked_to_stop_fast)
+				{
+					asked_to_stop_fast = 0;
+				}
```
This should be in shutdownNow check, otherwise we can swallow a SIGINT without doing anything about it. i.e. if the signal handler is triggered one line above this check.

DimCitus replied:
Oh. Right. Done now.

JelteF commented on src/bin/pg_autoctl/supervisor.c:
```diff
@@ -0,0 +1,880 @@
+						(void) supervisor_stop_subprocesses(supervisor);
+					}
+
+					(void) supervisor_shutdown_sequence(stoppingLoopCounter++);
```
This should be in a separate if, because `shutdownNow` is not always true anymore now when we are shutting down. We need either a separate boolean, e.g. `shuttingDown` or we reuse stoppingLoopCounter like `stoppingLoopCounter > 0 || shutdownNow`.

DimCitus replied:
Done now.

JelteF commented on src/bin/pg_autoctl/supervisor.c:
```diff
@@ -743,43 +741,31 @@ supervisor_may_restart(Service *service)
 				  service->name,
-				  restarts,
-				  (int) (now - oldestRestart));
+				  size,
```
I believe it should be this:
```suggestion
				  SUPERVISOR_SERVICE_MAX_TIME,
```

DimCitus replied:
I don't think so, because we allow up to MaxR restarts. We fail at MaxR+1, which is why the size of the ring buffer has to be MaxR+1 in the first place.

From https://erlang.org/doc/design_principles/sup_princ.html

> If more than MaxR number of restarts occur in the last MaxT seconds, the supervisor terminates all the child processes and then itself. The termination reason for the supervisor itself in that case will be shutdown.

JelteF replied:
I don't think we need ring buffer size of MaxR+1. We only store the current restart after checking the oldest. So we could overwrite that. Then we would still restart 5 times always and at the 6th time we consider if the entry we are about to overwrite is in our MaxT. If it is, we don't restart, because that would be restart MaxR+1 in MaxT seconds. If it is not, we can throw it out, because we do not need it anymore.

DimCitus replied:
Ah, yeah. Done that way now.

JelteF commented on src/bin/pg_autoctl/supervisor.c:
```diff
@@ -529,6 +469,77 @@ supervisor_stop(Supervisor *supervisor)
+	int signal = asked_to_stop_fast ? SIGINT : SIGTERM;
+	const char *signalStr = asked_to_stop_fast ? "SIGINT" : "SIGTERM";
+
+	if (asked_to_stop || asked_to_stop_fast)
```
nitpick: Inverting this check will allow for an early return. And the same can be done by adding an `else` case that the if/elseif chain that sets `shutdownNow`. That way the `shutdownNow` variable is not needed at all anymore.

DimCitus replied:
Early return makes it easier to read the code, done now.

That said we have more cases than meet the eyes and I think we still need the `shutdownNow` variable: if supervisor->shutdownSequenceInProgress and we received another SIGTERM (asked_to_stop_fast is false), then we don't want to signal all processes again.

Or maybe that's what you're saying, that we should signal our services each time we receive a signal, be it SIGINT or SIGTERM? Then your comment makes more sense, so that's what I have now pushed.

JelteF commented on src/bin/pg_autoctl/supervisor.c:
```diff
@@ -529,6 +469,77 @@ supervisor_stop(Supervisor *supervisor)
+		(void) supervisor_stop_subprocesses(supervisor);
+
+		/* allow for processing SIGINT again next time it's sent: reset */
+		if (asked_to_stop_fast)
```
I think there might still be a race condition if you send both SIGTERM and SIGINT very closely together. I think this should fix it, since we check which signal we handled this time, not which flags are set. 

```suggestion
		if (signal == SIGINT)
```

Otherwise `asked_to_stop_fast` might be set to true after setting `signal`. This way we do not unset the `asked_to_stop_flag` unless we actually handled it right now. If we did not handle it, it will simply be handled next loop, because it's still set.
